File: api/golang/core/lib/enclaves/kurtosis_mod.go
package enclaves

import (
	"github.com/go-yaml/yaml"
	"github.com/kurtosis-tech/stacktrace"
	"io/ioutil"
	"os"
)

const (
	dependenciesUrl = "https://docs.kurtosis.com/reference/starlark-reference/#dependencies"
)

type KurtosisMod struct {
	Module Module `yaml:"module"`
}

type Module struct {
	ModuleName string `yaml:"name"`
}

func parseKurtosisMod(kurtosisModFilepath string) (*KurtosisMod, error) {
	kurtosisModContents, err := ioutil.ReadFile(kurtosisModFilepath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, stacktrace.NewError("Couldn't find a '%v' in the root of the package at '%v'. Packages are expected to have a '%v' at root; have a look at '%v' for more", kurtosisYamlFilename, kurtosisModFilepath, kurtosisYamlFilename, dependenciesUrl)
		}
		return nil, stacktrace.Propagate(err, "An error occurred while reading the '%v' file at '%v'", kurtosisYamlFilename, kurtosisModFilepath)
	}

	var kurtosisModule KurtosisMod
	err = yaml.Unmarshal(kurtosisModContents, &kurtosisModule)
	if err != nil {
		return nil, stacktrace.Propagate(err, "An error occurred while parsing the '%v' file at '%v'", kurtosisYamlFilename, kurtosisModFilepath)
	}

	if kurtosisModule.Module.ModuleName == "" {
		return nil, stacktrace.NewError("Field module.name in %v needs to be set and cannot be empty", kurtosisYamlFilename)
	}

	return &kurtosisModule, nil
}

File: api/typescript/src/core/lib/enclaves/kurtosis_mod.ts
import * as fs from "fs"
import {err, ok, Result} from "neverthrow";
import * as yaml from "js-yaml";

export class KurtosisMod {
    constructor(
        public readonly module: Module,
    ){}
}

class Module {
    constructor(
        public readonly  name: string,
    ){}
}

const UTF8_ENCODING = "utf-8";

export async  function parseKurtosisMod(kurtosisModFilepath: string): Promise<Result<KurtosisMod, Error>> {
    // check if the mod file actually exists
    if (!fs.existsSync(kurtosisModFilepath)) {
        return err(new Error(`The file '${kurtosisModFilepath}' does not exist.`))
    }
    const kurtosisModFile = fs.readFileSync(kurtosisModFilepath, UTF8_ENCODING)
    const parsedYAML = (yaml.load(kurtosisModFile) as KurtosisMod)

    if (parsedYAML.module === null || parsedYAML.module.name === null || parsedYAML.module.name === "") {
        return err(new Error(`Field module.name in kurtosis.mod needs to be set and cannot be empty`))
    }

    return ok(parsedYAML)
}
